ortools/linear_solver/xpress_interface.cc

```diff
@@ -1931,9 +1999,104 @@ void XpressInterface::AddSolutionHintToOptimizer() {
     colind[i] = solver_->solution_hint_[i].first->index();
     val[i] = solver_->solution_hint_[i].second;
   }
-  if (int status = XPRSaddmipsol(mLp, len, val.get(), colind.get(), "USER_HINT")) {
-    LOG(WARNING) << "Failed to set solution hint.";
+  XPRSaddhint(mLp, len, val.get(), colind.get(), "USER_HINT");
+}
+
+void XpressInterface::SetCallback(MPCallback* mp_callback) {
+  if (callback_ != nullptr) {
+    // replace existing callback by removing it first
+    CHECK_STATUS(XPRSremovecboptnode(mLp, XpressOptNodeCallbackImpl, NULL));
+  }
+  callback_ = mp_callback;
+}
+
+// NOTE(user): This function must have this exact API, because we are passing
+// it to XPRESS as a callback.
+void XPRS_CC XpressOptNodeCallbackImpl(XPRSprob cbprob, void* cbdata, int* p_infeasible) {
+  auto callback_with_context = static_cast<const MPCallbackWithXpressContext*>(cbdata);
+  CHECK(callback_with_context != nullptr);
+  CHECK(callback_with_context->context != nullptr);
+  CHECK(callback_with_context->callback != nullptr);
+  if (callback_with_context->context->UpdateFromXpressState(cbprob)) {
+    // Only run user's callback if the context has changed
+    callback_with_context->callback->RunCallback(
```

Comment:
This is super risky: you are calling C++ code from a callstack that contains C code (the Xpress optimizer). If the C++ code raises an exception, then the stack will unwind even the C callstack and and up in the frame that called `XPRSmipoptimize()`. All cleanup implemented in the C code will be skipped and that will leave everything in an unexpected state.

What you should do here is:
1. Catch _any_ exception raised and store it somewhere.
2. In case of an exception call [XPRSinterrupt](https://www.fico.com/fico-xpress-optimization/docs/latest/solver/optimizer/HTML/XPRSinterrupt.html) to stop the solve.
3. After `XPRSmipoptimize()` returned, check whether an exception was raised and if so, re-raise it.

The documentation for [std::current_exception](https://en.cppreference.com/w/cpp/error/current_exception) has nice examples for this.

Reply:
I see your point. I started implementing what you suggested, but I realize that after the other changes I made, we don't really need to throw any exception during the call-back anymore.